src/seql/helpers.clj

```diff
@@ -217,21 +273,97 @@
           (qualify entity field)
           {:spec spec :handler handler}))
 
+(defmethod merge-entity-component :create-mutation
+  [{:keys [table entity] :as local-schema} {:keys [spec]}]
+  (let [mname   (qualify entity spec)
+        handler (fn [params]
+                  {:insert-into [table]
+                   :values      [(schema/as-row (m/schema params) entity params)]})]
+    (update local-schema :mutations assoc
+            mname
+            {:spec mname :handler handler})))
+
+(defmethod merge-entity-component :update-by-id-mutation
+  [{:keys [table entity] :as local-schema} {:keys [ident spec]}]
+  (let [mname   (qualify entity spec)
+        handler (fn [params]
+                  (let [id (get params ident)]
+                    {:update table
+                     :set    (schema/as-row (m/schema params) entity (dissoc params ident))
+                     :where  [:= (schema/resolve-field (m/schema params) ident) id]}))]
+    (update local-schema :mutations assoc
+            mname
+            {:spec mname :handler handler})))
+
+(defmethod merge-entity-component :delete-by-id-mutation
+  [{:keys [table entity] :as local-schema} {:keys [ident spec]}]
+  (let [mname   (qualify entity spec)
+        handler (fn [params]
+                  (let [id (get params ident)]
+                    {:delete-from table
+                     :where       [:= (schema/resolve-field (m/schema params) ident) id]}))]
+    (update local-schema :mutations assoc
+            mname
+            {:spec mname :handler handler})))
+
+(defmethod merge-entity-component :add-precondition
+  [{:keys [entity] :as local-schema} {:keys [handler key mutation]}]
+  (let [mname   (qualify entity mutation)]
+    (update-in local-schema [:mutations mname :pre] (comp vec conj) {:name key :query handler})))
+
+(defn- keys-spec-fields
+  "Makes use of `exoscale.coax.inspect/spec-root` to make a best effort
+   attempt at figuring out the keys present in a specific map-like spec.
+
+   Only keeps keys which share the provided spec."
+  [spec-name]
+  (letfn [(find-keys-flat [km]
+            (filter keyword? (flatten km)))
+          (find-keys [km]
+            (set (concat (find-keys-flat (:req km))
+                         (find-keys-flat (:opt km)))))
+          (key-set [spec-name]
+            (let [root (si/spec-root spec-name)]
+              (when (coll? root)
+                (let [[f & args :as _spec-form] root]
+                  (condp = f
+                    `s/keys       (find-keys (apply hash-map args))
+                    `s/merge      (into #{} (comp (keep key-set) cat) args)
+                    `s/multi-spec (key-set (s/form ((resolve (first args))))))))))]
+    (into []
```

Comment:
why not a set ? you could then ditch the distinct and that seems to make more sense, unless I am missing something.